lib/elixir/lib/calendar/duration.ex

```diff
@@ -283,4 +285,108 @@ defmodule Duration do
+  appear in any case, which is not the case in the base standard. `"P3M3W"` is not allowed in
+  ISO 8601-1, but it is in ISO 8601-2. Thus, any duration can be output by this function.
+
+  ## Examples
+
+      iex> Duration.new!(year: 3) |> Duration.to_iso8601()
+      "P3Y"
+      iex> Duration.new!(day: 28, hour: 6, minute: 42, second: 12) |> Duration.to_iso8601()
+      "P28DT6H42M12S"
+      iex> Duration.new!(second: 30) |> Duration.to_iso8601()
+      "PT30S"
+      iex> Duration.new!(second: 30) |> Duration.multiply(-1) |> Duration.to_iso8601()
+      "PT-30S"
+      iex> Duration.new!([]) |> Duration.to_iso8601()
+      "PT0S"
+
+      iex> Duration.new!(hour: 2, microsecond: {2, 6}) |> Duration.to_iso8601()
+      "PT2H0.000002S"
+      iex> Duration.new!(hour: 2, microsecond: {2, 3}) |> Duration.to_iso8601()
+      "PT2H0.000S"
+      iex> Duration.new!(hour: 2, microsecond: {2, 0}) |> Duration.to_iso8601()
+      "PT2H"
+      iex> Duration.new!(microsecond: {2, 0}) |> Duration.to_iso8601()
+      "PT0S"
+      iex> Duration.new!(microsecond: {2_000_000, 6}) |> Duration.to_iso8601()
+      "PT2S"
+  """
+
+  @spec to_iso8601(t) :: String.t()
+  def to_iso8601(duration)
+
+  def to_iso8601(%Duration{
+        year: 0,
+        month: 0,
+        week: 0,
+        day: 0,
+        hour: 0,
+        minute: 0,
+        second: 0,
+        microsecond: {0, _}
+      }) do
+    "PT0S"
+  end
+
+  def to_iso8601(%Duration{microsecond: {ms, 0}} = d) when ms != 0 do
+    to_iso8601(%Duration{d | microsecond: {0, 0}})
+  end
+
+  def to_iso8601(%Duration{} = d) do
+    "P#{to_iso8601_duration_date(d)}#{to_iso8601_duration_time(d)}"
+  end
+
+  defp to_iso8601_duration_date(d) do
+    year = unless d.year == 0, do: "#{d.year}Y"
+    month = unless d.month == 0, do: "#{d.month}M"
+    week = unless d.week == 0, do: "#{d.week}W"
+    day = unless d.day == 0, do: "#{d.day}D"
+
+    "#{year}#{month}#{week}#{day}"
+  end
+
+  defp to_iso8601_duration_time(%Duration{hour: 0, minute: 0, second: 0, microsecond: {0, _}}) do
+    ""
+  end
+
+  defp to_iso8601_duration_time(%Duration{microsecond: {ms, p}} = d)
+       when ms >= @microseconds_per_second do
+    second = d.second + div(ms, @microseconds_per_second)
+    microsecond = {rem(ms, @microseconds_per_second), p}
+
+    to_iso8601_duration_time(%Duration{d | second: second, microsecond: microsecond})
+  end
+
+  defp to_iso8601_duration_time(d) do
+    hour = unless d.hour == 0, do: "#{d.hour}H"
+    minute = unless d.minute == 0, do: "#{d.minute}M"
+
+    second =
+      case d do
+        %Duration{second: 0, microsecond: {0, _}} ->
+          ""
+
+        %Duration{microsecond: {0, _}} ->
+          "#{d.second}S"
+
+        %Duration{microsecond: {ms, p}} ->
+          microsecond =
+            ms
+            |> Integer.to_string()
+            |> String.pad_leading(6, "0")
+            |> binary_part(0, p)
+
+          "#{d.second}.#{microsecond}S"
```

Comment:
```
Duration.to_iso8601(%Duration{second: 1, microsecond: {-200000, 4}})
```

becomes `PT1.-200S` while it should be `PT-0.8000S`

Reply:
Good catch, thanks! I've revamped the implementation in https://github.com/elixir-lang/elixir/pull/13604/commits/2dd83341a30dab41dddfe60c2eff72a501ae36b6, and hopefully simplified it, based on what is done in `Calendar.ISO`. The difference with the latter is that there's no `1..999_999` limitation for microsecond in duration.

Now we always show `n` number according to the precision, which is consistent with `DateTime.to_iso8601/1` and others.